src/test/java/org/junitpioneer/jupiter/resource/ResourcesParallelismTests.java

```diff
@@ -0,0 +1,317 @@
+		}
+
+		@ParameterizedTest
+		@ValueSource(ints = { 1, 2, 3 })
+		void test3(@SuppressWarnings("unused") int iteration,
+				@SuppressWarnings("unused") @Shared(factory = TemporaryDirectory.class, name = SHARED_RESOURCE_C_NAME) Path directoryC,
+				@SuppressWarnings("unused") @Shared(factory = TemporaryDirectory.class, name = SHARED_RESOURCE_A_NAME) Path directoryA)
+				throws Exception {
+			failIfExecutedInParallel("test3Iteration" + iteration);
+		}
+
+	}
+
+	static class ThrowIfTestClassConstructorsRunInParallelTestCases1 {
+
+		ThrowIfTestClassConstructorsRunInParallelTestCases1(
+				// we don't actually use the resources, we just have them injected to verify whether sharing the
+				// same resources prevent the test constructors from running in parallel
+				@SuppressWarnings("unused") @Shared(factory = TemporaryDirectory.class, name = SHARED_RESOURCE_A_NAME, scope = GLOBAL) Path directoryA,
+				@SuppressWarnings("unused") @Shared(factory = TemporaryDirectory.class, name = SHARED_RESOURCE_B_NAME, scope = GLOBAL) Path directoryB)
+				throws Exception {
+			failIfExecutedInParallel("testConstructor1");
+		}
+
+		@Test
+		void fakeTest() {
+		}
+
+	}
+
+	static class ThrowIfTestClassConstructorsRunInParallelTestCases2 {
+
+		ThrowIfTestClassConstructorsRunInParallelTestCases2(
+				@SuppressWarnings("unused") @Shared(factory = TemporaryDirectory.class, name = SHARED_RESOURCE_B_NAME, scope = GLOBAL) Path directoryB,
+				@SuppressWarnings("unused") @Shared(factory = TemporaryDirectory.class, name = SHARED_RESOURCE_C_NAME, scope = GLOBAL) Path directoryC)
+				throws Exception {
+			failIfExecutedInParallel("testConstructor2");
+		}
+
+		@Test
+		void fakeTest() {
+		}
+
+	}
+
+	static class ThrowIfTestClassConstructorsRunInParallelTestCases3 {
+
+		ThrowIfTestClassConstructorsRunInParallelTestCases3(
+				@SuppressWarnings("unused") @Shared(factory = TemporaryDirectory.class, name = SHARED_RESOURCE_C_NAME, scope = GLOBAL) Path directoryC,
+				@SuppressWarnings("unused") @Shared(factory = TemporaryDirectory.class, name = SHARED_RESOURCE_A_NAME, scope = GLOBAL) Path directoryA)
+				throws Exception {
+			failIfExecutedInParallel("testConstructor3");
+		}
+
+		@Test
+		void fakeTest() {
+		}
+
+	}
+
+	// this method is written to fail if it is executed at overlapping times in different threads
+	private static void failIfExecutedInParallel(String testName) throws InterruptedException {
+		boolean wasZero = COUNTER.compareAndSet(0, 1);
+		assertThat(wasZero).isTrue();
+		// wait for the next test to catch up and potentially fail
+		Thread.sleep(TIMEOUT_MILLIS);
+		boolean wasOne = COUNTER.compareAndSet(1, 2);
+		assertThat(wasOne).isTrue();
+		// wait for the last test to catch up and potentially fail
+		Thread.sleep(TIMEOUT_MILLIS);
+		boolean wasTwo = COUNTER.compareAndSet(2, 0);
+		assertThat(wasTwo).isTrue();
+	}
+
+	// At time of writing, we couldn't find a way to test that @{Before, After}{All, Each}-annotated methods are not
```

Comment:
Isn't this promised by Jupiter? Is there anything we can do (wrong) to make these tests fail?

Reply:
Excellent question. I have no idea. :)

I remember reasoning that if a test class was run in parallel, it's `@BeforeEach` and `@AfterEach` methods could be called many times concurrently. Nothing I've found in the JUnit 5 User Guide seems to disprove this, but I probably haven't looked hard enough.

But I'm aware that the tests below may not be pulling their weight. So I'd be happy to remove this comment and the tests below if you agree.

Reply:
I looked into it and I think I understand what JUnit Platform does here. Each test (be it explicit, dynamic, or from a template) becomes a node in the test plan and that includes it's before/after behavior (see [this code](https://github.com/junit-team/junit5/blob/r5.9.0/junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTestTask.java#L139-L161) in Platform's `NodeTestTask`). Parallelization happens on the level of test plan nodes, which means multiple nodes can be executed at the same time. That, in turn, means that before/after extensions for _different_ nodes can run in parallel.

It should be possible to test this behavior by writing a test class that has a long-running before/after method and at least two tests. Then the 2+ tests become 2+ test nodes that get executed in parallel and so the before/after methods should run in parallel.

@jbduncan: I'll leave this explanation here and continue with the review. I may have time to try this out for myself, in which case I'll let you know here to avoid duplicate work.

Reply:
@nipafx Wow, thank you for looking so deeply into this! I'll see if I can test the before/after behaviour now as suggested, but I'm away on holiday from noon UK time until Monday morning, so if you don't hear back from me by the end of today, then feel free to have a go yourself.

Reply:
I'm having a go at this now. I'll let you know how it goes. :)

Reply:
Hmm, it's proving harder than I imagined to get some working code for or to reason through. Probably too late at night for me. :P

@nipafx If I revisit this and find myself equally stuck, would you prefer I keep the existing tests or remove them?